provider/pkg/provider/provider.go

```diff
@@ -466,6 +469,8 @@ func (k *commandProvider) Cancel(context.Context, *pbempty.Empty) (*pbempty.Empt
 }
 
 func (k *commandProvider) addContext(c context.Context) context.Context {
+	k.providerMutex.Lock()
+	defer k.providerMutex.Unlock()
```

Comment:
We should also add this wherever there's operations on the map - so also in the remove methon and cancel flow.

Reply:
Have added in 707450d

Reply:
Good addition